app/scripts/contentscript.js

```diff
@@ -276,3 +423,31 @@ function redirectToPhishingWarning(data = {}) {
   const querystring = new URLSearchParams({ hostname, href, newIssueUrl });
   window.location.href = `${baseUrl}#${querystring}`;
 }
+
+const initKeepWorkerAlive = () => {
+  setInterval(() => {
+    browser.runtime.sendMessage({ name: WORKER_KEEP_ALIVE_MESSAGE });
+  }, WORKER_KEEP_ALIVE_INTERVAL);
+};
+
+const start = () => {
+  const isDetectedPhishingSite =
+    window.location.origin === phishingPageUrl.origin &&
+    window.location.pathname === phishingPageUrl.pathname;
+
+  if (isDetectedPhishingSite) {
+    initPhishingStreams();
+    return;
+  }
+
+  if (shouldInjectProvider()) {
+    if (isManifestV3) {
+      initKeepWorkerAlive();
```

Comment:
It looks like this will keep the service worker alive regardless of whether the current website is connected or not. This runs for manifest V3 builds despite the connected status.

Reply:
To properly stop this interval when a site isn't connected would be challenging. The contentscript isn't aware of the connected status right now. Hmm.

We could spy on messages from here to detect connect/disconnect events though. Or we could update the provider to send special "connect"/"disconnect" JSON-RPC notifications to the contentscript just for this purpose.

Reply:
@Gudahtt 

I see. I confirmed your observations. Agreed we should look into starting this interval only when the site is connected in a separate PR as [you mentioned here](https://github.com/MetaMask/metamask-extension/pull/15494#pullrequestreview-1102798025). I created a new issue for this: https://github.com/MetaMask/metamask-extension/issues/15802

For clarification here, the service worker does terminate after some time when the pages themselves are closed. 

Repo steps to observe service worker terminating:
1. Inspect service workers (e.g. chrome://inspect/#service-workers)
2. Observe app-init.js service worker is not running. If it is running, be sure to close all `http://` or `https://` pages. Then, terminate the service worker.
3. Open dapp to run MM extension
4. Observe `app-init.js` service worker is running
5. Close dapp page and all other `http://` or `https://` pages
6. Wait about 5 minutes (current ETA for termination of inactive service worker)
7. Inspect service workers and observe `app-init.js` service worker is no longer running